### PR TITLE
Updated OC Interface HLD regarding rate utilization counters enhancements.

### PR DESCRIPTION
**Pull request by rvasanthm**
Changes:

1. Updated SONiC_OC_Interface_HLD.md doc with new counters attributes details and Interface_counters RPC enhancement.
2. New show command to display rate and utilization counters for Ethernet and port channel interfaces.
    show interface counters rate
3. show interface Ethernet <> and show interface PortChannel <> enhancements based on new counters and RPC enhancements..

### CODE REVIEW

File: doc/mgmt/SONiC_OC_Interface_HLD.md [hunk 1]
```diff
@@ -586,6 +648,31 @@ Mode of IPv4 address assignment: MANUAL
 IPv6 address is a::b/64
 Mode of IPv6 address assignment: MANUAL
 ```
+#### Display interface COUNTERS
+`show interface counters`
+Units: RX_MBPS/TX_MBPS(MB/s), RX_MbPS|TX_MbPS(Mb/s), RX_PPS|TX_PPS(pkts/s)
+
+Interface    State    RX_OK RX_MBPS  RX_MbPS  RX_PPS  RX_UTIL    RX_ERR    RX_DRP    RX_OVR    TX_OK  TX_MBPS    TX_MbPS  TX_PPS  TX_UTIL    TX_ERR    TX_DRP    TX_OVR
```
Comment (nnag-broadcom) [inline doc/mgmt/SONiC_OC_Interface_HLD.md]: RX_UTILS = will it show real time packets received in % ?

File: doc/mgmt/SONiC_OC_Interface_HLD.md [hunk 2]
```diff
@@ -586,6 +648,31 @@ Mode of IPv4 address assignment: MANUAL
...
+sonic# show interface Ethernet 64
+Ethernet64 is up, line protocol is up
+Hardware is Eth
+Mode of IPV4 address assignment: not-set
+Mode of IPV6 address assignment: not-set
+Interface IPv6 oper status: Disabled
+IP MTU 9100 bytes
+LineSpeed 25GB, Auto-negotiation off
+Last clearing of "show interface" counters: 1970-01-01 00:00:00
+30 seconds input rate 84640 bits/sec, 10236 Bytes/sec, 52 packets/sec
```
Comment (nnag-broadcom) [inline doc/mgmt/SONiC_OC_Interface_HLD.md]: Can we have packets/sec first followed by bits/sec?
30 seconds input rate 52 packets/sec, 84640 bits/sec, 10236 Bytes/sec

Generally most of the vendors follow this order and hence the ask

**Review by nnag-broadcom — COMMENTED** [review]
.

File: doc/mgmt/SONiC_OC_Interface_HLD.md (hunk 2 above)
Comment (rvasanthm) [reply doc/mgmt/SONiC_OC_Interface_HLD.md]: Addressed.

File: doc/mgmt/SONiC_OC_Interface_HLD.md (hunk 1 above)
Comment (rvasanthm) [reply doc/mgmt/SONiC_OC_Interface_HLD.md]: Utilization counters are calculated based on the configured load interval window. Back end calculates and populates  these counters (depends on load interval and polling interval) in counterdb and NBI just picks it up from there.
Please refer below HLD for more information:
https://github.com/BRCM-SONIC/sonic_doc_private/pull/43

**Comment by rupesh-k** [comment]
Don't see any mention about queue counters and yang changes and CLI changes.
Do I miss something ?

**Comment by rvasanthm** [comment]
Hello Rupesh,

CLI and Yang changes are not merged yet, PR’s opened and in WIP state, I am waiting to complete end to end validation with your backend changes. Please push your changes.

PR details:


  1.  https://gitlab.force10networks.com/e-sonic/sonic-mgmt-common/merge_requests/219
  2.  https://gitlab.force10networks.com/e-sonic/sonic-mgmt-framework/merge_requests/328
  3.  https://gitlab.force10networks.com/e-sonic/sonic-spytest/merge_requests/233

-Ravi

**Comment by rupesh-k** [comment]
Thanks Ravi, I didn't see any mention about Queue Counters here in this document. So my question was regarding that.
I hope u are covering  that in your implementation.

Sure. I will update u once I push my changes.
I dont have access to these gitlab links. 
May be I can pick from your branch in github, so I can do integration testing in my side once I push my changes.
Can I reach u on Slack for quick collaboration in next week.

Thanks
Rupesh Kumar

**Comment by rvasanthm** [comment]
Hello Rupesh,

Queue counters changes was done by Ashok D (added him to the mail thread) and changes are already merged.
This HLD covers only interface rate and utilization counters enhancements.

-Ravi

**Comment by rvasanthm** [comment]
Hi,

While supporting Klish “show queue counters” based on attached mail, I added queue rate counters in Klish “show queue counters” CLI.
If back end added  below 2 counters in DB then we will see result in Klish CLI>

    "SAI_QUEUE_STAT_PACKETS_PER_SECOND": "0",

    "SAI_QUEUE_STAT_BYTES": "0",

    "SAI_QUEUE_STAT_BYTES_PER_SECOND": "0",

Output in Klish:

sonic# show queue counters
----------------------------------------------------------------------------------------------------------------------------------------------
Port        TxQ  Counter/pkts          Counter/bytes         Rate/PPS              Rate/BPS              Drop/pkts             Drop/bytes
----------------------------------------------------------------------------------------------------------------------------------------------
CPU         MC0  0                     0                     0/s                   0/s                   0                     0
CPU         MC1  0                     0                     0/s                   0/s                   0                     0
CPU         MC2  0                     0                     0/s                   0/s                   0                     0

Regards
Ashok D
From: Vasanthm, Ravi <Ravi_Vasanthm@Dell.com>
Sent: Monday, July 27, 2020 10:31 AM
To: project-arlo/SONiC; project-arlo/SONiC
Cc: Author; Daparthi, Ashok
Subject: RE: [project-arlo/SONiC] Updated OC Interface HLD regarding rate utilization counters enhancements. (#102)

Hello Rupesh,

Queue counters changes was done by Ashok D (added him to the mail thread) and changes are already merged.
This HLD covers only interface rate and utilization counters enhancements.

-Ravi

**Pull request merged**